rrule/src/parser/parsers.rs

```diff
@@ -33,44 +24,51 @@ pub(crate) struct ParsedDateStringTime {
     pub sec: u32,
 }
 
-fn get_datetime_captures<T: FromStr>(
-    captures: &Captures,
-    idx: usize,
-    val: &str,
-) -> Result<T, ParseError> {
-    captures
-        .get(idx)
-        .ok_or_else(|| ParseError::InvalidDateTimeFormat(val.into()))?
-        .as_str()
-        .parse()
-        .map_err(|_| ParseError::InvalidDateTimeFormat(val.into()))
-}
-
 impl ParsedDateString {
     /// Parses a date string with format `YYYYMMDD(THHMMSSZ)` where the part in parentheses
     /// is optional. It returns [`ParsedDateString`].
     pub(crate) fn from_ical_datetime(val: &str) -> Result<Self, ParseError> {
-        let captures = DATESTR_RE
-            .captures(val)
-            .ok_or_else(|| ParseError::InvalidDateTimeFormat(val.into()))?;
-
-        let year = get_datetime_captures(&captures, 1, val)?;
-        let month = get_datetime_captures(&captures, 2, val)?;
-        let day = get_datetime_captures(&captures, 3, val)?;
-
-        // Check if time part is captured
-        let time = if captures.get(4).is_some() {
-            let hour = get_datetime_captures(&captures, 5, val)?;
-            let min = get_datetime_captures(&captures, 6, val)?;
-            let sec = get_datetime_captures(&captures, 7, val)?;
-            Some(ParsedDateStringTime { hour, min, sec })
-        } else {
-            None
-        };
+        let len = val.find(|c| c == '\n' || c == '\r').unwrap_or(val.len());
+        if len < 8 {
+            // Not a valid YYYYMMDD date.
+            return Err(ParseError::InvalidDateTimeFormat(val.into()));
+        }
 
-        let zulu_timezone_set = match captures.get(8) {
-            Some(part) => part.as_str() == "Z",
-            None => false,
+        // Parse date (YYYYMMDD).
+        let year = val[0..4]
```

Comment:
This will panic if the character at byte 3 is a multi-byte character.

Such an input would be absolute garbage, but the parser should not crash in such a case and return an error instead.

Reply:
Given that valid RRULEs are strictly ASCII, you can probably treat this as a raw byte array instead of a string.

Reply:
> This will panic if the character at byte 3 is a multi-byte character.

That's very true, thanks for catching! And I agree that it should be handled gracefully.

And yes, either ensuring that string is ascii or using `.get(..)` instead of `[..]` would probably solve this. I will add some tests and a fix.

Reply:
> Given that valid RRULEs are strictly ASCII, you can probably treat this as a raw byte array instead of a string.

Btw, raw byte arrays (`&[u8]`) doesn't have the convenient parsing methods for integers so that's why I didn't convert it but just checked that it's all ascii.